src/index.ts

```diff
@@ -139,3 +139,99 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
     return { puts, nextCounter }
   }
 }
+
+/**
+ * Update a history table with an auto-incrementing attribute value in DynamoDB
+ *
+ * @example
+ * ```
+ * import { DynamoDB } from '@aws-sdk/client-dynamodb'
+ * import { DynamoDBDocument } from '@aws-sdk/lib-dynamodb'
+ * import { DynamoDBHistoryAutoIncrement } from '@nasa-gcn/dynamodb-autoincrement'
+ *
+ * const client = new DynamoDB({})
+ * const doc = DynamoDBDocument.from(client)
+ *
+ * const autoIncrementHistory = DynamoDBHistoryAutoIncrement({
+ *   doc,
+ *   counterTableName: 'widgets', // The table storing the current item
+ *   counterTableKey: {
+ *     widgetID: 42 // ID of the item to be updated
+ *   },
+ *   attributeName: 'version',
+ *   tableName: 'widgetsHistory', // The table storing the history of items in
+ *   initialValue: 1,
+ * })
+ *
+ * const latestVersionNumber = await autoIncrementHistory.put({
+ *   widgetName: 'A new name for this item',
+ *   costDollars: 199.99,
+ * })
+ * ```
+ */
+export class DynamoDBHistoryAutoIncrement extends BaseDynamoDBAutoIncrement {
+  protected async next(item: Record<string, NativeAttributeValue>) {
+    let nextCounter
+
+    const existingEntry = (
+      await this.props.doc.get({
+        TableName: this.props.counterTableName,
+        Key: this.props.counterTableKey,
+      })
+    ).Item
+
+    const counter: number | undefined =
+      existingEntry?.[this.props.attributeName]
+
+    let untrackedEntryPutCommandInput: PutCommandInput | undefined = undefined
+    if (counter === undefined) {
+      nextCounter = existingEntry
+        ? this.props.initialValue + 1
+        : this.props.initialValue
+    } else {
+      nextCounter = counter + 1
+    }
+
+    if (counter === undefined && existingEntry) {
+      untrackedEntryPutCommandInput = {
+        TableName: this.props.tableName,
+        Item: {
+          ...existingEntry,
+          [this.props.attributeName]: this.props.initialValue,
+        },
+      }
+    }
+
+    const Item = {
+      ...item,
+      ...this.props.counterTableKey,
+      [this.props.attributeName]: nextCounter,
+    }
+
+    const puts: PutCommandInput[] = [
+      {
+        ConditionExpression: 'attribute_not_exists(#counter)',
+        ExpressionAttributeNames: {
+          '#counter': this.props.attributeName,
+        },
+        Item,
+        TableName: this.props.tableName,
+      },
+      {
+        ConditionExpression:
+          'attribute_not_exists(#counter) OR #counter <> :counter',
+        ExpressionAttributeNames: {
+          '#counter': this.props.attributeName,
+        },
+        ExpressionAttributeValues: {
+          ':counter': nextCounter,
+        },
+        Item,
+        TableName: this.props.counterTableName,
+      },
+    ]
+    if (untrackedEntryPutCommandInput) puts.push(untrackedEntryPutCommandInput)
```

Comment:
I think we agreed that we would only store the _current_ value in counterTable, and only _past_ values in table.

```suggestion
    const existingItem = (
      await this.props.doc.get({
        TableName: this.props.counterTableName,
        Key: this.props.counterTableKey,
      })
    ).Item

    const counter: number | undefined =
      existingItem?.[this.props.attributeName]

    if (counter === undefined) {
      nextCounter = this.props.initialValue

      // Existing item didn't have a version, so give it one
      if (existingItem) {
        existingItem[this.props.attributeName] = nextCounter
        nextCounter += 1
      }
    } else {
      nextCounter = counter + 1
    }

    const puts: PutCommandInput[] = [
      {
        ConditionExpression: 'attribute_not_exists(#counter)',
        ExpressionAttributeNames: {
          '#counter': this.props.attributeName,
        },
        Item: existingItem,
        TableName: this.props.tableName,
      },
      {
        ConditionExpression:
          'attribute_not_exists(#counter) OR #counter <> :counter',
        ExpressionAttributeNames: {
          '#counter': this.props.attributeName,
        },
        ExpressionAttributeValues: {
          ':counter': nextCounter,
        },
        Item: {
          ...item,
          ...this.props.counterTableKey,
          [this.props.attributeName]: nextCounter,
        },
        TableName: this.props.counterTableName,
      },
    ]
```

Reply:
Ah yea you are right, I will need to update the tests and make sure this is behaving as expected